### PR TITLE
split r-hat

### PR DESCRIPTION
**Pull request by bob-carpenter**
I added 

* implementation of split-R-hat (with tests)
* additional validation in basic R-hat to ensure each chain has 2 or more draws (I added an additional test for this)

This resolves issue #36.

### CODE REVIEW

**Comment by bob-carpenter** [comment]
I can update for 100% coverage of the tests.  And apply `black`.

File: bayes_kit/rhat.py [hunk 1]
```diff
@@ -40,3 +42,45 @@ def rhat(chains: list[SeqType]) -> FloatType:
...
+    [[1, 2], [3, 4], [5, 6], [7]]
+    ```
+
+    Parameters:
+    chains: list of univariate Markov chains
+
+    Returns:
+    List of input chains split in half.
+    """
+    return [list(arr) for chain in chains for arr in np.array_split(chain, 2)]
```
Comment (WardBrian) [inline bayes_kit/rhat.py:62]: I don't think it's necessary to cast `arr` back into a list, is it?

File: test/test_rhat.py [hunk 2]
```diff
@@ -30,14 +30,33 @@ def test_rhat_ragged():
...
+def test_rhat_size_exceptions():
+    bad1 = []
+    bad2 = [[1.01, 1.2, 1.3, 1.4]],
+    bad3 = [[1, 2, 3], [4]]
+    for chains in [bad1, bad2, bad3]:
+        with pt.raises(ValueError):
+            rhat(chains)
+
+def test_split_chains():
+    print("hello")
```
Comment (WardBrian) [inline test/test_rhat.py:45]:
```suggestion

```

**Review by WardBrian — COMMENTED** [review]
A few minor comments

**Comment by bob-carpenter** [comment]
This is ready to review.  Failures are all in MCMC tests, not in the rhat stuff I added.

This is all updated with 100% coverage tests.  I just added split_rhat and rank_normalized_rhat to the same rhat.py file.  The doc has full definitions and pointers to the original references.

File: bayes_kit/rhat.py [hunk 3]
```diff
@@ -1,11 +1,88 @@
...
+    [[1, 2], [3, 4], [5, 6], [7]]
+    ```
+
+    Parameters:
+    chains: List of univariate Markov chains.
+
+    Returns:
+    List of input chains split in half.
+    """
+    return [list(arr) for chain in chains for arr in np.array_split(chain, 2)]
```
Comment (WardBrian) [inline bayes_kit/rhat.py:25]: I don't believe it is necessary to cast `arr` back to a `list`

File: bayes_kit/rhat.py [hunk 4]
```diff
@@ -1,11 +1,88 @@
...
+
+
+def rank_chains(chains: list[SeqType]) -> list[SeqType]:
+    """
+    Returns a copy of the included Markov chains transformed with
+    ranks normalized to [0, 1] and an offset inverse CDF. The ranks
```
Comment (WardBrian) [inline bayes_kit/rhat.py:31]: I can sort-of back it out from the definition of the function, but I think it would be useful to give more immediate definitions of what terms like "ranks" mean here. In particular it seems like the returned values of this function are not in `[0, 1]`, so I'm not sure what normalization means here

File: bayes_kit/rhat.py [hunk 5]
```diff
@@ -1,11 +1,88 @@
...
+
+
+def rank_normalize_chains(chains: list[SeqType]) -> list[SeqType]:
+    """Return the rank-normalized version of the input chains.
+
+    The rank-normalized value for element `j` of list `i` is
+    ```
+    inverse_Phi((rank[i][j] - 3/8) / (size(chains) - 1/4),
+    ```
+    where `inv_Phi` is the inverse cumulative distribution function for
```
Comment (WardBrian) [inline bayes_kit/rhat.py:60]: We should be consistent between `inverse_Phi` and `inv_Phi`

File: bayes_kit/rhat.py [hunk 6]
```diff
@@ -1,11 +1,88 @@
...
+
+    Returns:
+    List of chains with values replaced by rank-normalized values.
+    """
+    S = sum([len(chain) for chain in chains])
+    result = chains
+    ranked_chains = rank_chains(chains)
+    for i, chain_i in enumerate(ranked_chains):
+        for j, rank_ij in enumerate(chain_i):
+            val = sp.stats.norm.ppf((rank_ij - 0.325) / (S - 0.25))
```
Comment (WardBrian) [inline bayes_kit/rhat.py:81]: I just wanted to note that this is adding a new dependency - previously, scipy was used _only_ in testing

File: bayes_kit/rhat.py [hunk 7]
```diff
@@ -27,10 +104,13 @@ def rhat(chains: list[SeqType]) -> FloatType:
...
 
     Throws:
-    ValueError: if there are fewer than two chains
+    ValueError: If there is not at least one chain or if any chain has
+    fewer than two elements.
     """
     if len(chains) < 2:
         raise ValueError(f"rhat requires len(chains) >= 2, but {len(chains) = }")
+    if not all(len(chain) >= 2 for chain in chains):
+        raise ValueError(f"rhat requires len(chain) >= 2 for all chain in chains")
```
Comment (WardBrian) [inline bayes_kit/rhat.py:113]:
```suggestion
        raise ValueError(f"rhat requires len(chain) >= 2 for every chain in chains")
```

File: bayes_kit/rhat.py [hunk 8]
```diff
@@ -27,10 +104,13 @@ def rhat(chains: list[SeqType]) -> FloatType:
     R-hat statistic
 
     Throws:
-    ValueError: if there are fewer than two chains
+    ValueError: If there is not at least one chain or if any chain has
+    fewer than two elements.
```
Comment (WardBrian) [inline bayes_kit/rhat.py:108]:
```suggestion
    ValueError: If there is not at least one chain.
    ValueError: If any chain has fewer than two elements.
```

Typically different raise statements are documented separately like this rather than by saying "or"

File: bayes_kit/rhat.py [hunk 9]
```diff
@@ -40,3 +120,61 @@ def rhat(chains: list[SeqType]) -> FloatType:
...
+
+    See :func:`split_chains` for a specification of splitting.
+
+    Parameters:
+    chains: List of univariate Markov chains.
+
+    Returns:
+    Split R-hat statistic.
+
+    Throws:
```
Comment (WardBrian) [inline bayes_kit/rhat.py:147]: Same as above

File: bayes_kit/rhat.py [hunk 10]
```diff
@@ -40,3 +120,61 @@ def rhat(chains: list[SeqType]) -> FloatType:
...
+    See :func:`split_rhat` for a specification of split R-hat and
+    :func:`rank_normalize_chains` for rank normalization.
+
+    Parameters:
+    chains: List of univariate Markov chains.
+
+    Returns:
+    Rank-normalized R-hat statistic.
+
+    Throws:
```
Comment (WardBrian) [inline bayes_kit/rhat.py:176]: Same as above

File: bayes_kit/rhat.py [hunk 11]
```diff
@@ -1,11 +1,88 @@
...
+    """
+    Return a list of the input chains split in half.  The result will be
+    a list twice as long as the input.  For example,
+    ```
+    >>> split_chains([[1, 2, 3, 4], [5, 6, 7]])
+    [[1, 2], [3, 4], [5, 6], [7]]
+    ```
+
+    Parameters:
+    chains: List of univariate Markov chains.
```
Comment (WardBrian) [inline bayes_kit/rhat.py:20]: This is true for all the docstrings in this file, and is difficult to really enforce until we have something like Sphinx building the docs to let us know, but these docstrings are not properly formatted Google style. 

The correct version would look like
```suggestion
    Args:
        chains: List of univariate Markov chains.
```

File: test/test_rhat.py [hunk 12]
```diff
@@ -33,11 +41,117 @@ def test_rhat_ragged():
...
-        chain1 = [1.01, 1.2, 1.3, 1.4]
-        chains = [chain1]
-        rhat(chains)
+def test_rhat_size_exceptions():
+    bad1 = []
+    bad2 = ([[1.01, 1.2, 1.3, 1.4]],)
+    bad3 = [[1, 2, 3], [4], [5, 6, 7, 8, 9]]
+    for chains in [bad1, bad2, bad3]:
+        with pt.raises(ValueError):
+            rhat(chains)
```
Comment (WardBrian) [inline test/test_rhat.py:50]: This style of test should be written to use [`pytest.mark.parameterize`](https://docs.pytest.org/en/7.3.x/how-to/parametrize.html)

```suggestion
@pt.mark.parametrize("chains", [[],  ([[1.01, 1.2, 1.3, 1.4]],),  [[1, 2, 3], [4], [5, 6, 7, 8, 9]]])
def test_rhat_size_exceptions(chains):
    with pt.raises(ValueError):
        rhat(chains)
```

File: test/test_rhat.py [hunk 13]
```diff
@@ -33,11 +41,117 @@ def test_rhat_ragged():
...
+    np.testing.assert_allclose(
+        rhat([[1, 2, 2], [3, 4, 3]]), split_rhat([[1, 2, 2, 3, 4, 3]])
+    )
+    np.testing.assert_allclose(
+        rhat([[1, -2, 3], [4, 5, 6], [7, 8], [9, 12]]),
+        split_rhat([[1, -2, 3, 4, 5, 6], [7, 8, 9, 12]]),
+    )
+
+
+def test_split_rhat_size_exceptions():
```
Comment (WardBrian) [inline test/test_rhat.py:75]: Same as above, use `pt.mark.parametrize`

File: test/test_rhat.py [hunk 14]
```diff
@@ -33,11 +41,117 @@ def test_rhat_ragged():
...
+    rn14 = rank_norm(1, 4)
+    rn24 = rank_norm(2, 4)
+    rn34 = rank_norm(3, 4)
+    rn44 = rank_norm(4, 4)
+    np.testing.assert_allclose(
+        [[rn34, rn24], [rn14, rn44]], rank_normalize_chains([[3.9, 3.1], [2.2, 5.9]])
+    )
+
+
+def test_rank_normalized_rhat_size_exceptions():
```
Comment (WardBrian) [inline test/test_rhat.py:127]: Same as above

File: test/test_rhat.py
Comment (WardBrian) [inline test/test_rhat.py:1]: In general, I think some of the new tests (specifically for the rhat functions, not the helper functions) should be commented a bit to explain what each assert is verifying or where the 'magic numbers' came from

**Review by WardBrian — CHANGES_REQUESTED** [review]

File: bayes_kit/rhat.py [hunk 15]
```diff
@@ -40,3 +120,61 @@ def rhat(chains: list[SeqType]) -> FloatType:
...
+    return rhat(split_chains(chains))
+
+
+def rank_normalized_rhat(chains: list[SeqType]) -> FloatType:
+    """Return the rank-normalized R-hat for the specified chains.
+    Rank normalized r-hat replaces each value in the chains with its
+    rank, applies a shifted inverse standard normal cdf, and
+    returns the split R-hat value of the result.
+
+    Rank-normalized R-hat was introduced in the following paper.
```
Comment (jsoules) [inline bayes_kit/rhat.py:160]: Might be good to include a sentence or two about why you'd want to do this instead of normal r-hat. (I know there's a paper cite, but that's a pretty big detour.)

File: test/test_rhat.py (hunk 12 above)
Comment (jsoules) [reply test/test_rhat.py:50]: Syntax permitting, I'd further suggest that this is easier to read if you give each example on its own line & add a comment about what it's testing ("at least two chains required", "at least two chains required", "Every chain must have at least two elements")

File: test/test_rhat.py [hunk 16]
```diff
@@ -33,11 +41,117 @@ def test_rhat_ragged():
...
+
+
+def test_split_chains():
+    np.testing.assert_equal([], split_chains([]))
+    np.testing.assert_equal([[1], []], split_chains([[1]]))
+    np.testing.assert_equal([[1], [2]], split_chains([[1, 2]]))
+    np.testing.assert_equal([[1, 2], [3]], split_chains([[1, 2, 3]]))
+    np.testing.assert_equal([[1, 2, 3], [4, 5, 6]], split_chains([[1, 2, 3, 4, 5, 6]]))
+    np.testing.assert_equal(
+        [[1, 2], [3], [4, 5], [6, 7]], split_chains([[1, 2, 3], [4, 5, 6, 7]])
```
Comment (jsoules) [inline test/test_rhat.py:60]: Why's it decide 3 has to sit by itself? (The example in the docstring said it'd be 7 that's isolated.)

Is there some variation in implementations or something?

File: test/test_rhat.py
Comment (jsoules) [reply test/test_rhat.py:1]: Agreed

File: bayes_kit/rhat.py [hunk 17]
```diff
@@ -1,11 +1,88 @@
...
+    List of chains with values replaced by transformed ranks.
+    """
+    if len(chains) == 0:
+        return chains
+    flattened = np.concatenate(chains)
+    ranks = flattened.argsort().argsort() + 1
+    reshaped_arrays = []
+    current_index = 0
+    for array in chains:
+        size = len(array)
```
[the remark this answers is not in the document]
Comment (WardBrian) [reply bayes_kit/rhat.py:47]: ArrayLike contains all sorts of things you can’t call len() on, like `int`. For most of our purposes we probably shouldn’t be using it, though these won’t really be caught until #26

File: bayes_kit/rhat.py (hunk 3 above)
Comment (bob-carpenter) [reply bayes_kit/rhat.py:25]: @WardBrian was right---I didn't need it as the internal function was able to proceed without the the `list()` cast.

File: bayes_kit/rhat.py (hunk 4 above)
Comment (bob-carpenter) [reply bayes_kit/rhat.py:31]: @jsoules was correct forensically.  I updated the docs to match what they're currently doing and included examples.

File: bayes_kit/rhat.py (hunk 5 above)
Comment (bob-carpenter) [reply bayes_kit/rhat.py:60]: settled on `inv_Phi`.

File: bayes_kit/rhat.py (hunk 6 above)
Comment (bob-carpenter) [reply bayes_kit/rhat.py:81]: Thanks for the reminder.  I updated the README to list the dependency.  

I can't figure out a way around using SciPy---it contains a lot of special functions like inverse CDFs that aren't in NumPy.

File: bayes_kit/rhat.py (hunk 8 above)
Comment (bob-carpenter) [reply bayes_kit/rhat.py:108]: Fixed.  Thanks for letting me know.

File: bayes_kit/rhat.py (hunk 10 above)
Comment (bob-carpenter) [reply bayes_kit/rhat.py:176]: same fix.

File: bayes_kit/rhat.py (hunk 11 above)
Comment (bob-carpenter) [reply bayes_kit/rhat.py:20]: fixed.

File: test/test_rhat.py (hunk 12 above)
Comment (bob-carpenter) [reply test/test_rhat.py:50]: I find the `parametrize` [sic] syntax really clunky but made the changes as suggested.   I find the misspelling particularly annoying.

File: test/test_rhat.py
Comment (bob-carpenter) [reply test/test_rhat.py:1]: Could you be more specific about what you'd like me to comment?

The rhat function tests rely on a brute-force definition from  `test_rhat.py` and then from manually splitting.  I don't see anywhere to add more doc, but am happy to add more if there's something that's confusing.

File: bayes_kit/rhat.py [hunk 18]
```diff
@@ -1,11 +1,88 @@
...
+    Parameters:
+    chains: list of univariate Markov chains
+
+    Returns:
+    List of chains with values replaced by transformed ranks.
+    """
+    if len(chains) == 0:
+        return chains
+    flattened = np.concatenate(chains)
+    ranks = flattened.argsort().argsort() + 1
```
[the remark this answers is not in the document]
Comment (bob-carpenter) [reply bayes_kit/rhat.py:43]: Tests fail with only a single .argsort().  This is a bit worrying in that I asked ChatGPT to write this function and I wrote exhaustive tests, but I clearly don't understand it well enough. 

 Here's an example to show why we need two applications of `.argsort()`:

```python
>>> chains = [[4.2, 5.7], [7.2, 6.1], [-12.9, 107]]

>>> flattened = np.concatenate(chains)

>>> flattened
array([  4.2,   5.7,   7.2,   6.1, -12.9, 107. ])

>>> flattened.argsort()
array([4, 0, 1, 3, 2, 5])

>>> flattened.argsort().argsort()
array([1, 2, 4, 3, 0, 5])
```

To help, the doc notes that if you have an array and argsort and then apply, you get something in sorted order.

```python
>>> flattened[flattened.argsort()]
array([-12.9,   4.2,   5.7,   6.1,   7.2, 107. ])
```

So the first `.argsort()` makes sense.  But then I'm not sure why we don't get a fixed point after one application or why that works with the rest of the code.

Here's the doc:  https://numpy.org/doc/stable/reference/generated/numpy.argsort.html

The `+1` is because we're indexing from 1 in the transform.

File: bayes_kit/rhat.py [hunk 19]
```diff
@@ -1,11 +1,88 @@
...
+        current_index += size
+    return reshaped_arrays
+
+
+def rank_normalize_chains(chains: list[SeqType]) -> list[SeqType]:
+    """Return the rank-normalized version of the input chains.
+
+    The rank-normalized value for element `j` of list `i` is
+    ```
+    inverse_Phi((rank[i][j] - 3/8) / (size(chains) - 1/4),
```
[the remark this answers is not in the document]
Comment (bob-carpenter) [reply bayes_kit/rhat.py:58]: I'd argue the doc string is so simple that the code looks just like it.    I added a citation to the original 1958 paper, which is where the constants come from and a short note as to what the constants are for.  The other functions already have citations.

File: bayes_kit/rhat.py [hunk 20]
```diff
@@ -1,11 +1,88 @@
...
+    For a specification of ranking, see :func:`rank_chains`.
+
+    Parameters:
+    chains: List of univariate Markov chains.
+
+    Returns:
+    List of chains with values replaced by rank-normalized values.
+    """
+    S = sum([len(chain) for chain in chains])
+    result = chains
```
[the remark this answers is not in the document]
Comment (bob-carpenter) [reply bayes_kit/rhat.py:77]: > There's no need for a full deep-copy here since you aren't actually preserving any of the pre-exisitng values. Could either preallocate len(ranked_chains) arrays?

Could you maybe suggest how I do that?  I don't know how to preallocate or what you're suggesting I do with it.

I also don't know what the `[[]] * 5` example is supposed to be warning me about.

I do understand copying vs. reference, just not how it applies in Python.

File: bayes_kit/rhat.py (hunk 9 above)
Comment (bob-carpenter) [reply bayes_kit/rhat.py:147]: fixed.

File: bayes_kit/rhat.py (hunk 7 above)
Comment (bob-carpenter) [reply bayes_kit/rhat.py:113]: fixed.

File: test/test_rhat.py (hunk 13 above)
Comment (bob-carpenter) [reply test/test_rhat.py:75]: done.

File: test/test_rhat.py (hunk 14 above)
Comment (bob-carpenter) [reply test/test_rhat.py:127]: also done.

File: bayes_kit/rhat.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (bob-carpenter) [reply bayes_kit/rhat.py:47]: Any suggestion on what I should be using for typing instead?

File: bayes_kit/rhat.py [hunk 21]
```diff
@@ -1,11 +1,88 @@
...
+    ```
+    inverse_Phi((rank[i][j] - 3/8) / (size(chains) - 1/4),
+    ```
+    where `inv_Phi` is the inverse cumulative distribution function for
+    the standard normal distribution and
+    ```
+    rank[i][j] = rank_chains(chains)[i][j]
+    ```
+    is the rank of element `i` in chain `j` and `size(chains)` is the
+    total number of elements in the chains.
```
[the remark this answers is not in the document]
Comment (bob-carpenter) [reply bayes_kit/rhat.py:66]: There's no list anywhere, so I don't see what to do with bullets.

File: bayes_kit/rhat.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (jsoules) [reply bayes_kit/rhat.py:77]: > Could you maybe suggest how I [preallocate arrays]?

Performance-wise, it may not make too much of a difference to pre-allocate vs. just doing this:
```python
result = []

for i, chain_i in enumerate(ranked_chains):
    result.append([]) # add a new list to result variable
    for j, rank_ij in enumerate(chain_i):
        val = sp.stats.norm.ppf((rank_ij - 0.325) / (S - 0.25))
        result[i].append(val)
return result
```
assuming the `i`s and `j`s are well-behaved and always start at 0 & increment, which they should here.

It depends how much data there is whether constantly allocating new elements will actually be noticeable. (Python list allocations start with 4 by default and double every time you reach the end of allocated space, but they're pretty quick about it, so the calls to `append` may not actually hurt you noticeably).

Since the chains may have different lengths, I think the best you could do would be:

```python
result = list(None for n in range(len(ranked_chains)))
for i, chain_i in enumerate(ranked_chains):
    result[i] = list(None for n in range(len(chain_i)))
    for j, rank_ij in enumerate(chain_i):
        val = sp.stats.norm.ppf((rank_ij - 0.325) / (S - 0.25))
        result[i][j] = val
return result
```
It is also possible to do this as a one-liner:
```python
result = list(list(None for x in range(len(chain_i))) for chain_i in ranked_chains)
```
but I can't imagine there would be a performance improvement that would justify writing something so hard to parse. You'd be better off just using `numpy` with a predefined size if it actually mattered:

```python
result = np.empty((n,m), dtype='uint16')
```

assuming you have a square array and don't mind getting uninitialized memory (`np.empty`); can use `np.zeros` the same way if you need it zeroed out.

---
Regarding the `[[]] * 5` example--Python has a builtin to allocate lists of known length:
```python
In [14]: foo = [5] * 10

In [15]: foo
Out[15]: [5, 5, 5, 5, 5, 5, 5, 5, 5, 5]
```
The problem is if you preallocate a list of objects this way--including a list of empty lists--it will allocate each element of the list you're constructing to point to the *same* instance of the object. So if you try to build a list of lists as `bar = [[]] * 5` thinking to get 5 empty lists, you actually get 5 pointers to the *same* empty list, with predictably hilarious results when you try to operate on them individually and wind up modifying them all, every time. (I honestly don't know why that functionality is even still in the language spec.)

File: bayes_kit/rhat.py [hunk 22]
```diff
@@ -15,7 +92,7 @@ def rhat(chains: list[SeqType]) -> FloatType:
     `phi[m] = mean(chains[m])` and `psi[m] = var(chains[m])`.  This reduces to
     the standard definition when all chains are the same length.
```
[the remark this answers is not in the document]
Comment (bob-carpenter) [reply bayes_kit/rhat.py:93]: My bad.  I meant to write `chains` where I wrote `theta` (the variable `theta` is what's used in all the original papers and doc).  

I removed "standard" and just cited the paper and noted that the paper only defines rhat for rectangular structures whereas this definition generalizes to ragged structures in a way that matches the original for rectangular inputs.

Is it that you don't know the `mean` function or that it's confusing that `N` is an array or that we're talking about the mean of a sample rather than a random variable?  I clarified all three points, so it's now extra pedantic. I'm sticking to `N` because that matches the paper we're implementing and the books that define it.   I also clarified we're using sample mean and sample variance (as opposed to random variable mean and variance) in case that was confusing you and supplied the NumPy functions to compute the exact values (hope that's not too close to the code).  I'd strongly prefer not to define sample mean and variance in the definition of rhat.

The actual interpretation of the value is very tricky.  It's about the scale (aka std deviation) of the values in the chains.  (So it's not the scale of the current estimate that's reduced, but the scale of the estimator---estimators are random variables and thus have a standard deviation---usually the sample standard deviation of an estimator is called the "standard error").

I can't cite BDA3, because the definition there is different and could be confusing, and it's also bad form to cite secondary sources.  I do cite BDA3 and the Stan Reference Manual for split R-hat as that's where it was first described.

File: bayes_kit/rhat.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (jsoules) [reply bayes_kit/rhat.py:58]: Adding the reference is good, but I would suggest--first, jumping out of the context of this code to an outside reference can be a pretty distracting or train-of-thought-breaking thing when you just want to understand the  code (to the point that many readers may not even attempt it); and second, the paper is likely to provide an explanation that's in a register that may be difficult to parse for the casual reader (by which I guess I mean, for anybody who didn't immediately understand what was happening from the code itself). I'd really be looking just for a bit more intuition here, like explaining what rank-normalization is ("projecting ranks so that they're evenly distributed on a normal curve", if that's correct, etc).

**Comment by bob-carpenter** [comment]
@jsoules and @WardBrian.  Thanks for all the comments.  I'll learn Python one PR at a time.

I made all the fixes I could understand and marked them as resolved.  

There are a couple points I didn't know how to fix and I would really appreciate guidance.

File: bayes_kit/rhat.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (bob-carpenter) [reply bayes_kit/rhat.py:53]: I updated my comment to explain why it works, but our comments crossed paths.  @roualdes sorted me out and I understand what it's doing now.  Apparently this is a standard trick.  I'm a bit reluctant to include stackoverflow links in my doc.

File: bayes_kit/rhat.py [hunk 23]
```diff
@@ -1,11 +1,88 @@
...
+    Parameters:
+    chains: List of univariate Markov chains.
+
+    Returns:
+    List of input chains split in half.
+    """
+    return [list(arr) for chain in chains for arr in np.array_split(chain, 2)]
+
+
+def rank_chains(chains: list[SeqType]) -> list[SeqType]:
```
[the remark this answers is not in the document]
Comment (bob-carpenter) [reply bayes_kit/rhat.py:28]: I think I am going to need more help on the types here.  I can't piece together yours and @WardBrian's comments into actions.

File: bayes_kit/rhat.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (bob-carpenter) [reply bayes_kit/rhat.py]: When C++ documents cosine, they don't write an intro to trig in the doc.  It says, "Computes the cosine of num (measured in radians)."  It assumes if you want to use trig functions that you know trig.  I would like to make roughly the same kind of decision with our doc.  

Inverse cdfs are standard, sample means are standard, and sample variance is standard, in very much the same way as cosine is standard.

The rank normalization isn't so standard that everyone knows it, but it uses a very simple function to map to (0, 1) evenly, then it uses the well-known normal inverse CDF to project back to the real line.  That's all there is to it.  If the reader doesn't know how inverse CDFs work, I don't think the doc to the `rank_normalize_chains` is the place for a tutorial.  

I wasn't sure what you mean by "evenly distributed on a normal curve" but the spacing will not be even when you project back through the inverse CDF.  That is, `inv_Phi([0.1, 0.2, ..., 0.9])` does not have even spaced values even though `[0.1, ..., 0.9]` does.  Of course, calling `inv_Phi` (a standard function) `sp.stats.norm.ppf` isn't helping anyone.


```python
>>> y = (np.arange(10) + 0.5) / 10

>>> y
array([0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95])

>>> sp.stats.norm.ppf(y)
array([-1.64485363, -1.03643339, -0.67448975, -0.38532047, -0.12566135,
        0.12566135,  0.38532047,  0.67448975,  1.03643339,  1.64485363])
```

[the remark this answers is not in the document]
Comment (bob-carpenter) [reply bayes_kit/rhat.py]: P.S.  3/8 and 1/4 is to make sure the results are in (0, 1).  The exact values don't matter so much other than that you don't want to get too close to 0 or 1 or that pushes you close to -inf, +inf.   3/8 and 1/4 are just a reasonable compromise in keeping the values reasonable for various sizes of inputs.

File: test/test_rhat.py (hunk 16 above)
Comment (jsoules) [reply test/test_rhat.py:65]: Ah, okay! But what about the `[[1, 2, 3], [4, 5, 6, 7]]` example? Ultimately this is an implementation detail that doesn't really matter, and probably doesn't need to be documented; I'm just curious.

File: test/test_rhat.py (hunk 12 above)
Comment (WardBrian) [reply test/test_rhat.py]: Apparently it is an acceptable 'alternative' spelling in American english, though I also wish there was another e.

The primary alternative to `parametrize` is just splitting the tests into different functions, which has the nice benefit that the function names can serve as those comments @jsoules recommended.

In case it wasn't clear, the benefit of doing this versus the for loop approach is specificity. When you use `parametrize`, pytest reports each different parameter value as a unique "test" in the output, so if one of them fails you can immediately see which one. In the for-loop approach, if any of them fail you just get notified that the entire test failed.

File: bayes_kit/rhat.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (WardBrian) [reply bayes_kit/rhat.py:57]: It's a judgement call, but I'm in favor of using our own types instead of `ArrayLike` for things like this. `Union[typing.Sequence[float], npt.NDArray[np.float64]]` is probably the minimal type we actually will pass for a chain.

Writing out our own type has some downsides (like we might need to broaden it over time if we also wanted to pass e.g. `list[int]`) but is useful if only because we can avoid things being in it that we don't want.

If we want to stick with `npt.ArrayLike`, you essentially just need to insert `np.asarray` or `np.asanyarray` calls everywhere before doing anything, since `ArrayLike` is really only defined as "things you can pass to `np.asarray`"

File: bayes_kit/rhat.py (hunk 3 above)
Comment (WardBrian) [reply bayes_kit/rhat.py:25]: @bob-carpenter you marked this as resolved but the code did not actually get changed

File: test/test_rhat.py (hunk 14 above)
Comment (WardBrian) [reply test/test_rhat.py:133]: @bob-carpenter this code also did not get updated/pushed it seems

File: bayes_kit/rhat.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (jsoules) [reply bayes_kit/rhat.py:28]: I suspect @WardBrian knows the numpy type system better than I do, so I'll defer to him on the typing question.

Upon further reflection, since the chains may not be of the same length, this probably has to stay a `list` of something, so my earlier comment was irrelevant.

File: test/test_rhat.py (hunk 16 above)
Comment (bob-carpenter) [reply test/test_rhat.py:65]: I added `[[1, 2, 3], [4, 5, 6, 7]]` as a test case and added it as an example in the doc.

This implementation detail matters in the sense that the answer will be different with two different splits.

File: bayes_kit/rhat.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (jsoules) [reply bayes_kit/rhat.py]: I don't think there's a need to write an intro to stats in the doc here; but "rank normalization" doesn't even have a Wikipedia entry & it's not "normalization" in the most obvious sense (of converting the values to a 0-1 range).

I think I'd be looking for something like "Rank normalization maps the ranks to the range (0, 1) and then returns the quantiles of the standard-normal distribution for the resulting values. (A small margin is first applied to avoid infinite values from the ppf function). \n\n Specifically, the rank-normalized value ..." This is pretty much the information that's present in the doc for the `rank_normalized_rhat` function, but when you're looking at this function, you probably don't realize you ought to look over there for it.

File: bayes_kit/rhat.py (hunk 15 above)
Comment (bob-carpenter) [reply bayes_kit/rhat.py:209]: I added a one-sentence explanation for each, but I don't know how to say anything meaningful without a lot of stats jargon (mean, variance, distribution, stationary distribution, Markov chain, normality, trend, Cauchy distribution).

>  Rank-normalized R-hat should be more robust in situations where the stationary distribution of the Markov chains is not normal (e.g., for density targets for which means and/or variances are undefined, such as the Cauchy distribution).

> The main utility of splitting is to diagnose non-stationary chains  (e.g., ones with an upward or downward trend).  Unlike the base `rhat(chains)` function, this version is applicable to a single Markov chain.

Comment (jsoules) [reply bayes_kit/rhat.py:209]: That seems clear to me!

File: bayes_kit/rhat.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (bob-carpenter) [reply bayes_kit/rhat.py:57]: I went with @WardBrian's suggestion for now. We may eventually want to go with something that'd work with single precision, but then we need to be careful throughout not to upcast.

@WardBrian also pointed out that the function will still work with `np.float32` input, it just won't pass mypy checks.

[the remark this answers is not in the document]
Comment (bob-carpenter) [reply bayes_kit/rhat.py:57]: I added a more specific typing as @WardBrian suggested elsewhere.

File: bayes_kit/rhat.py (hunk 3 above)
Comment (bob-carpenter) [reply bayes_kit/rhat.py:25]: Committed change with removed cast.

**Comment by bob-carpenter** [comment]
> A few minor comments

27 of them and counting.  :-(.

There are two remaining concerns that I do not know how to address:

* https://github.com/flatironinstitute/bayes-kit/pull/37#discussion_r1248101521
* https://github.com/flatironinstitute/bayes-kit/pull/37#discussion_r1247958390

File: bayes_kit/rhat.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (bob-carpenter) [reply bayes_kit/rhat.py:102]: OK, used a comprehension.

File: test/test_rhat.py
Comment (bob-carpenter) [reply test/test_rhat.py:1]: added comments

File: bayes_kit/rhat.py [hunk 24]
```diff
@@ -1,36 +1,158 @@
...
+    where 
+    * `inv_Phi` is the inverse cumulative distribution function for
+    the standard normal distribution,
+    * `rank[i][j] = rank_chains(chains)[i][j]` is the rank of element
+    `i` in chain `j`, and 
+    * `size(chains)` is the total number of elements in the chains.
+
+    For a specification of ranking, see :func:`rank_chains`.
+
+    The transformed values will be int he same order as the original
```
Comment (WardBrian) [inline bayes_kit/rhat.py:77]:
```suggestion
    The transformed values will be in the same order as the original
```

**Review by WardBrian — APPROVED** [review]
I found one typo, but I promise this is my last comment. Feel free to merge after fixing

File: bayes_kit/rhat.py (hunk 24 above)
Comment (bob-carpenter) [reply bayes_kit/rhat.py:77]: fixed.

File: bayes_kit/rhat.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (bob-carpenter) [reply bayes_kit/rhat.py]: I just copied in what you suggested (but without the parens).   So as not to have redundant doc, I then erased where I said something similar two paragraphs later.  

The definition of "normalization" being used here is in the general sense of

https://en.wikipedia.org/wiki/Normalization_(statistics)

and is very close to the specific sense of

https://en.wikipedia.org/wiki/Quantile_normalization

with a standard Gaussian distribution and the specific offsets introduced in the cited paper.
